In [86]:
from IPython.display import HTML

HTML(
    """
    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap" rel="stylesheet">

    <link rel="stylesheet" href="custom.css">
    <script src="custom.js" defer></script>
    <style>
        div.jp-OutputPrompt.jp-OutputArea-prompt {
            display: none !important;
        }
    
        .jp-OutputPrompt {
            display: none !important;
        }

        .jp-Notebook-cell.celltag_remove_cell {
            display: none !important;
        }
    </style>

    """
)

In [87]:
from IPython.display import HTML, Image, display


html_header = """

<header style="position: relative; overflow: hidden; background: #ffffff; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Inter, Roboto, Arial, sans-serif;">
  <div style="position: absolute; inset: 0; z-index: 0;" aria-hidden="true">
    <svg viewBox="0 0 1400 160" preserveAspectRatio="none" xmlns="http://www.w3.org/2000/svg" style="width:100%; height:100%; display:block;">
      <g stroke="#0c447c" stroke-opacity="0.06" stroke-width="1">
        <line x1="0" y1="40" x2="1400" y2="40"/><line x1="0" y1="80" x2="1400" y2="80"/><line x1="0" y1="120" x2="1400" y2="120"/>
      </g>
      <g fill="#7b828e" fill-opacity="0.28">
        <rect x="120" y="70" width="6" height="64"/><rect x="160" y="55" width="6" height="88"/><rect x="200" y="60" width="6" height="20"/><rect x="240" y="40" width="6" height="55"/><rect x="280" y="65" width="6" height="30"/><rect x="320" y="45" width="6" height="40"/><rect x="360" y="72" width="6" height="22"/><rect x="880" y="58" width="6" height="42"/><rect x="920" y="66" width="6" height="26"/><rect x="960" y="42" width="6" height="50"/><rect x="1000" y="60" width="6" height="32"/><rect x="1040" y="50" width="6" height="44"/><rect x="1080" y="68" width="6" height="24"/><rect x="1120" y="55" width="6" height="38"/>
      </g>
      <g stroke="#7b828e" stroke-opacity="0.28" stroke-width="1">
        <line x1="123" y1="60" x2="123" y2="110"/><line x1="163" y1="48" x2="163" y2="108"/><line x1="203" y1="52" x2="203" y2="86"/><line x1="243" y1="32" x2="243" y2="100"/><line x1="283" y1="58" x2="283" y2="100"/><line x1="323" y1="38" x2="323" y2="90"/><line x1="363" y1="65" x2="363" y2="100"/><line x1="883" y1="50" x2="883" y2="104"/><line x1="923" y1="58" x2="923" y2="96"/><line x1="963" y1="34" x2="963" y2="96"/><line x1="1003" y1="52" x2="1003" y2="96"/><line x1="1043" y1="42" x2="1043" y2="98"/><line x1="1083" y1="60" x2="1083" y2="96"/><line x1="1123" y1="46" x2="1123" y2="97"/>
      </g>
      <polyline fill="none" stroke="#5a6270" stroke-opacity="0.35" stroke-width="1.5"
        points="0,120 60,110 140,95 220,100 300,75 380,85 460,60 540,68 620,45 700,58 780,38 860,50 940,30 1020,42 1100,25 1180,35 1260,18 1400,28"/>
    </svg>
  </div>
  <div style="position: relative; z-index: 1; max-width: 1200px; margin: 0 auto; padding: 18px 24px; display: flex; align-items: center; justify-content: space-between; gap: 24px;">
    <div style="display: flex; align-items: center; gap: 12px;">
      <svg width="30" height="60" viewBox="0 0 30 30" xmlns="http://www.w3.org/2000/svg" role="img" aria-label="Database icon" style="flex-shrink:0;">
        <ellipse cx="15" cy="6" rx="11" ry="4" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <path d="M4 6 V15 C4 17.2 9 19 15 19 C21 19 26 17.2 26 15 V6" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <path d="M4 15 V24 C4 26.2 9 28 15 28 C21 28 26 26.2 26 24 V15" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <ellipse cx="15" cy="15" rx="11" ry="4" fill="none" stroke="#7b828e" stroke-width="1.4"/>
      </svg>
    </div>
  </div>
</header>

"""

display(HTML(html_header))


In [88]:
from IPython.display import HTML, Image, display
from datetime import date

today_str = date.today().strftime("%B %d, %Y")

html_coverpage_container = f"""
<div id="cover-page-container">
    <div id="title-container">
        <h1>Indicators</h1>
        <div id="meta-data-container">
            <div>{today_str}</div>
            <div>by Aaron Hardy</div>
        </div>
    </div>
    <div id="title-bar"></div>
</div>
"""


display(HTML(html_coverpage_container))

# Summary

This project involves an end-to-end process of gathering historical stock fundamentals data and generating a stock trading strategy from it using SQL and python. The raw data was sourced from [SimFin.com](https://www.simfin.com/en/fundamental-data-download/), which you will need a free account to download. MySQL was used as the database management system.

## Load Data
The raw data existed as CSVs. Each CSV was loaded into a MySQL table. The code used to load each table can be seen [here](load_data.html)

In [89]:
USE fundamentals_data;

SyntaxError: invalid syntax (1622834613.py, line 1)

In [ ]:
SELECT * 
FROM industries 
ORDER BY IndustryId ASC 
LIMIT 10;

In [ ]:

from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT *
FROM industries 
ORDER BY IndustryId ASC 
LIMIT 10;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
df = df.map(lambda x: x.strip() if isinstance(x, str) else x)
HTML(df.to_html(index=False))

IndustryId,Industry,Sector
100001,Industrial Products,Industrials
100002,Business Services,Industrials
100003,Engineering & Construction,Industrials
100004,Waste Management,Industrials
100005,Industrial Distribution,Industrials
100006,Airlines,Industrials
100007,Consulting & Outsourcing,Industrials
100008,Aerospace & Defense,Industrials
100009,Farm & Construction Machinery,Industrials
100010,Transportation & Logistics,Industrials


In [ ]:
SELECT * 
FROM us_balance_quarterly
LIMIT 10;

In [ ]:

from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT * 
FROM us_balance_quarterly
LIMIT 10;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.to_html(index=False))

Ticker,SimFinId,Currency,FiscalYear,FiscalPeriod,ReportDate,PublishDate,RestatedDate,SharesBasic,SharesDiluted,CashAndEquiv,AccountsNotesReceivable,Inventories,TotalCurrentAssets,PropertyPlantEquipmentNet,LongTermInvestmentsReceivables,OtherLongTermAssets,TotalNoncurrentAssets,TotalAssets,PayablesAccruals,ShortTermDebt,TotalCurrentLiabilities,LongTermDebt,TotalNoncurrentLiabilities,TotalLiabilities,ShareCapitalAdditionalPaidIn,TreasuryStock,RetainedEarnings,TotalEquity,TotalLiabilitiesEquity
A,45846,USD,2020,Q3,2020-07-31,2020-09-01,2020-09-01,309000000,312000000,1358000000,930000000,746000000,3245000000,846000000,148000000,5307000000,6301000000,9546000000,311000000,40000000,1314000000,2283000000,3251000000,4565000000,5327000000,0,130000000,4981000000,9546000000
A,45846,USD,2020,Q4,2020-10-31,2020-12-18,2021-12-17,308000000,311000000,1441000000,1038000000,720000000,3415000000,845000000,158000000,5209000000,6212000000,9627000000,639000000,75000000,1467000000,2284000000,3287000000,4754000000,5314000000,0,81000000,4873000000,9627000000
A,45846,USD,2021,Q1,2021-01-31,2021-03-02,2021-03-02,306000000,309000000,1329000000,1087000000,755000000,3483000000,866000000,165000000,5160000000,6191000000,9674000000,656000000,314000000,1687000000,2185000000,3183000000,4870000000,5269000000,0,4000000,4804000000,9674000000
A,45846,USD,2021,Q2,2021-04-30,2021-06-01,2021-06-01,306000000,306000000,1380000000,1075000000,791000000,3514000000,884000000,188000000,5812000000,6884000000,10398000000,738000000,205000000,1758000000,2727000000,3830000000,5588000000,5274000000,0,-12000000,4810000000,10398000000
A,45846,USD,2021,Q3,2021-07-31,2021-09-01,2021-09-01,303000000,306000000,1428000000,1122000000,818000000,3632000000,905000000,204000000,5750000000,6859000000,10491000000,728000000,130000000,1724000000,2728000000,3821000000,5545000000,5310000000,0,90000000,4946000000,10491000000
A,45846,USD,2021,Q4,2021-10-31,2021-12-17,2022-12-21,305000000,307000000,1575000000,1172000000,830000000,3799000000,945000000,185000000,5776000000,6906000000,10705000000,774000000,0,1708000000,2729000000,3608000000,5316000000,5323000000,0,348000000,5389000000,10705000000
A,45846,USD,2022,Q1,2022-01-31,2022-03-03,2022-03-03,301000000,303000000,1158000000,1205000000,879000000,3474000000,974000000,191000000,5688000000,6853000000,10327000000,801000000,0,1584000000,2730000000,3589000000,5173000000,5293000000,0,159000000,5154000000,10327000000
A,45846,USD,2022,Q2,2022-04-30,2022-05-31,2022-05-31,299000000,301000000,1207000000,1237000000,937000000,3643000000,1010000000,190000000,5612000000,6812000000,10455000000,780000000,175000000,1813000000,2730000000,3520000000,5333000000,5295000000,0,160000000,5122000000,10455000000
A,45846,USD,2022,Q3,2022-07-31,2022-09-01,2023-03-03,298000000,299000000,1053000000,1405000000,1038000000,3778000000,1100000000,195000000,5459000000,6754000000,10532000000,909000000,36000000,1861000000,2733000000,3366000000,5227000000,5328000000,0,324000000,5305000000,10532000000
A,45846,USD,2022,Q4,2022-10-31,2022-12-21,2023-12-20,318000000,325000000,1053000000,1405000000,1038000000,3778000000,1100000000,195000000,5459000000,6754000000,10532000000,909000000,36000000,1861000000,2733000000,3366000000,5227000000,5328000000,0,324000000,5305000000,10532000000


The list of U.S. companies Excel file was inspected in Excel and the file contained rows of data split over several cells in an unpredictable way. In Excel, the CONCAT() function was used to convert all rows to a single concatenated string which was then split again using Data > Text to Columns tool.

Next, we uploaded each SimFin CSV file into a separate SQL database table. The example below shows how the table `us_companies` was uploaded to the MySQL database titled 'fundamentals_data'. All SimFin data files were uploaded similarly.

In [ ]:

DROP TABLE IF EXISTS us_companies;

-- Create the table structure
CREATE TABLE us_companies (
    Ticker VARCHAR(255),
    SimFinId INT,
    CompanyName VARCHAR(255),
    IndustryId INT,
    ISIN VARCHAR(255),
    EndOfFinancialYearMonth INT,
    NumberEmployees INT,
    BusinessSummary TEXT,
    Market VARCHAR(255),
    CIK INT,
    MainCurrency VARCHAR(255)
);

-- Upload the csv data
LOAD DATA LOCAL INFILE 'c:\\Users\\aaron\\Desktop\\Local Projects\\Portfolio Projects\\data\\SimFin\\us-companies\\us-companies.csv'
INTO TABLE us_companies
FIELDS TERMINATED BY ','
OPTIONALLY ENCLOSED BY '"'
LINES TERMINATED BY '\n'
IGNORE 1 LINES; -- skip header row


The company profile table (`us_companies`) had 37 rows where both the Ticker and Company name were NULL.

In [ ]:
SELECT * FROM us_companies;

UPDATE us_companies
SET `Company Name` = NULL, `Ticker` = NULL
WHERE `Company Name` = '' AND `Ticker` = '';

SELECT COUNT(*) 
FROM us_companies
WHERE (Ticker IS NULL AND `Company Name` IS NULL);


Next, we deleted from the table any rows that had a NULL value for `Ticker` and `CompanyName`. These rows had insufficient identifying information.

In [ ]:

DELETE FROM us_companies 
WHERE (Ticker IS NULL AND CompanyName IS NULL);

The data includes several financial statement tables. The range of dates for all balance sheet, income statement, and cash flow tables were confirmed below.

In [ ]:

SELECT 
    "us_balance_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_quarterly
UNION ALL
SELECT 
    "us_balance_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_annual
UNION ALL
SELECT 
    "us_balance_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_ttm
UNION ALL
SELECT 
    "us_income_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_quarterly
UNION ALL
SELECT 
    "us_income_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_annual
UNION ALL
SELECT 
    "us_income_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_ttm
UNION ALL
SELECT 
    "us_cashflow_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_quarterly
UNION ALL
SELECT 
    "us_cashflow_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_annual
UNION ALL
SELECT 
    "us_cashflow_ttm" AS table_name,        
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_ttm;





In [ ]:
import json
import mysql.connector

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

cursor = conn.cursor()
cursor.execute(f"ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '{creds['password']}';")
cursor.execute("FLUSH PRIVILEGES;")

In [ ]:

from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT 
    "us_balance_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_quarterly
UNION ALL
SELECT 
    "us_balance_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_annual
UNION ALL
SELECT 
    "us_balance_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_ttm
UNION ALL
SELECT 
    "us_income_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_quarterly
UNION ALL
SELECT 
    "us_income_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_annual
UNION ALL
SELECT 
    "us_income_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_ttm
UNION ALL
SELECT 
    "us_cashflow_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_quarterly
UNION ALL
SELECT 
    "us_cashflow_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_annual
UNION ALL
SELECT 
    "us_cashflow_ttm" AS table_name,        
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_ttm;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.to_html(index=False))

table_name,min_date,max_date,years
us_balance_quarterly,2020-07-31,2025-05-31,4.8323
us_balance_annual,2020-07-31,2025-05-31,4.8323
us_balance_ttm,2020-07-31,2025-05-31,4.8323
us_income_quarterly,2020-07-31,2025-05-31,4.8323
us_income_annual,2020-07-31,2025-05-31,4.8323
us_income_ttm,2020-07-31,2025-05-31,4.8323
us_cashflow_quarterly,2020-07-31,2025-05-31,4.8323
us_cashflow_annual,2020-07-31,2025-05-31,4.8323
us_cashflow_ttm,2020-07-31,2025-05-31,4.8323


The data spans a period starting *July 31, 2020* and ending *May 31, 2025*, covering about 4.8 years.

Next, we look at the stock price table (`us_shareprices_monthly`) and 

In [ ]:

-- Clean daily prices
SELECT 
    MIN(Close),
    MAX(Close),
    MIN(AdjClose),
    MAX(AdjClose)
FROM us_shareprices_daily;

In [90]:

from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT 
    MIN(Close),
    MAX(Close),
    MIN(AdjClose),
    MAX(AdjClose)
FROM us_shareprices_daily;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
HTML(df.to_html(index=False))

MIN(Close),MAX(Close),MIN(AdjClose),MAX(AdjClose)
0.0000,99999999.9900,0.0000,99999999.9900


The `Close` price ranged from \\$0 to \\$99,999,999.99. 
The `Adj. Close` price ranged from \\$0 to \\$9,984,240.00.

We removed tickers where a `Close` or `AdjClose` price for any day exceeded \\$100,000, since the common investor is not likely to have an investing budget exceeding \\$100,000 for any single ticker.

In [ ]:
SELECT DISTINCT Ticker
    FROM us_shareprices_daily 
    WHERE Close > 100000 OR AdjClose > 100000;

In [98]:
from IPython.display import HTML, display
import json
import mysql.connector
import pandas as pd

with open("mysql_creds.json") as f:
    creds = json.load(f)

conn = mysql.connector.connect(
    host=creds["host"],
    user=creds["user"],
    password=creds["password"],
    database=creds["database"],
    auth_plugin="mysql_native_password"
)

query = """
SELECT DISTINCT Ticker
    FROM us_shareprices_daily 
    WHERE Close > 100000 OR AdjClose > 100000;
"""

cursor = conn.cursor()
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
csv_string = ", ".join(df["Ticker"].astype(str))
print (f"{len(df)} tickers:")
print(csv_string)



31 tickers:
ALLR, APDN, APVO, BCAN, BRK-A, CETX, CJET, CYN, EJH, ERNA, EZGO, FFIE, GDHG, HCTI, HLBZ, HOLO, HSDT, HUBC, JAGX, KXIN, MCMJ, MULN, PIXY, SHPH, TGL, TNXP, VCIG, VLCN, WHLR, WKHS, XXII


In [ ]:
DROP TABLE IF EXISTS us_shareprices_daily_filtered;

CREATE TABLE us_shareprices_daily_filtered AS
SELECT *
FROM us_shareprices_daily
WHERE
    Ticker NOT IN(
        SELECT DISTINCT
            Ticker
        FROM us_shareprices_daily
        WHERE
            Close > 100000
            OR AdjClose > 100000
    )

SELECT * FROM us_shareprices_daily_filtered LIMIT 10;

WITH full_table AS (
        SELECT *, ROW_NUMBER() OVER (
                PARTITION BY
                    Ticker
                ORDER BY Date DESC
            ) AS rn
        FROM us_shareprices_daily_filtered
)
SELECT *
FROM full_table
WHERE rn <= 3
LIMIT 9

In [107]:
from IPython.display import HTML 
import json
import pandas as pd
from sqlalchemy import create_engine

with open("mysql_creds.json") as f:
    creds = json.load(f)

engine = create_engine(
    f"mysql+pymysql://{creds['user']}:{creds['password']}@localhost/{creds['database']}")


query = """
WITH full_table AS (
        SELECT *, ROW_NUMBER() OVER (
                PARTITION BY
                    Ticker
                ORDER BY Date DESC
            ) AS rn
        FROM us_shareprices_daily_filtered
)
SELECT *
FROM full_table
WHERE rn <= 3
LIMIT 9
"""

df = pd.read_sql(query, engine)
HTML(df.to_html(index=False))

# -- df = pd.DataFrame(cursor.fetchall(), columns=[desc[0] for desc in cursor.description])
# -- HTML(df.to_html(index=False))

Ticker,SimFinId,Date,Open,High,Low,Close,AdjClose,Volume,Dividend,SharesOutstanding,rn
,18728220,2025-06-20,12.19,13.37,11.49,12.17,12.17,167628,0.0,15674052,1
,18728220,2025-06-18,9.16,12.00,9.15,11.96,11.96,84360,0.0,15674052,2
,18728220,2025-06-17,9.01,10.14,9.01,9.84,9.84,85917,0.0,15674052,3
A,45846,2025-06-20,116.17,116.86,115.08,115.56,114.65,3043869,0.0,284064728,1
A,45846,2025-06-18,116.09,116.88,114.67,115.52,114.61,1560338,0.0,284064728,2
A,45846,2025-06-17,117.50,118.47,115.82,116.09,115.17,1577141,0.0,284064728,3
AA,367153,2025-06-20,28.53,28.61,27.83,27.86,27.60,7894057,0.0,258902338,1
AA,367153,2025-06-18,28.63,28.85,28.13,28.35,28.09,5634805,0.0,258902338,2
AA,367153,2025-06-17,29.05,29.49,28.61,28.72,28.46,5188049,0.0,258902338,3


## Metrics

In [ ]:
SELECT
    cfq.Ticker,
    cfq.`Report Date`,
    cfq.`Net Cash from Operating Activities`,
    isq.`Net Income`,
    isq.`Operating Income (Loss)`
    -- isq.`Net Income`/cfq.`Net Cash from Operating Activities` AS CashFlowQuality
FROM
    us_cashflow_quarterly cfq
    LEFT JOIN us_income_quarterly isq ON cfq.Ticker = isq.Ticker
    AND cfq.`Report Date` = isq.`Report Date`
    -- WHERE isq.`Net Income`/cfq.`Net Cash from Operating Activities` IS NOT NULL
UNION

SELECT
    cfq.Ticker,
    cfq.`Report Date`,
    cfq.`Net Cash from Operating Activities`,
    isq.`Net Income`,
    isq.`Operating Income (Loss)`
FROM
    us_cashflow_quarterly cfq
    RIGHT JOIN us_income_quarterly isq ON cfq.Ticker = isq.Ticker
    AND cfq.`Report Date` = isq.`Report Date`
LIMIT 10;

In [ ]:

-- WHERE rn = (
--     SELECT 
--         Ticker,
--         ROW_NUMBER() OVER (PARTITION BY Ticker ORDER BY Date ASC) AS rn 
-- )
-- where `Adj. Close` > 1E6;


-- --------------
-- MONTHLY PRICES
-- --------------

drop table if exists us_shareprices_monthly;
-- When you drop the table, all indexes disappear automatically, so you must recreate them.

CREATE TABLE us_shareprices_monthly AS
SELECT
    Ticker,
    MAX(Date) AS Date,
    -- GROUP_CONCAT() → builds a list of Close prices
	-- ORDER BY Date DESC → newest price first
	-- SUBSTRING_INDEX(..., ',', 1) → pick the first price
    SUBSTRING_INDEX(
        GROUP_CONCAT(Close ORDER BY Date DESC),
        ',',
        1
    ) AS Close,
    SUBSTRING_INDEX(
        GROUP_CONCAT(`Adj. Close` ORDER BY Date DESC),
        ',',
        1
    ) AS AdjClose
FROM us_shareprices_daily
GROUP BY
    Ticker,
    YEAR(Date),
    MONTH(Date);
    
    
CREATE INDEX idx_month_end_ticker_date
ON us_shareprices_monthly (Ticker, Date);
